Acquiring locks are deferred to commit time.

burqen opened the pull request:
Introducing setting _deferred_locking_, off by default.

We separate between optimistic and pessimistic locking:

**Optimistic locking**
Acquire locks triggered by some write or read event through the `Statement` cake.
Instead of grabbing those locks we only remind what resources we want to lock and then acquire the them right before commit. 

**Pessimistic locking**
Locks grabbed explicitly by the user or grabbed on master by a slave. Those locks are taken immediately. 

**Note**
Grabbing locks just before commit guards for store inconsistencies but not logical inconsistencies. 

**Warning**
This feature should not be used. Most of all code in Neo4j is assumed to be executed under some sort of lock, exclusive or shared. This commit makes that assumption false and there has not been enough investigation to foresee what consequences that will have for transaction execution. These problems are still in the unknown but there are also known, purposely undocumented, problems.

Commit:
Prototype "finished", no tests, can maybe compile, no run, to be continued

Commit:
Added DeferringLocks and added naive and stupid what to bypass them
in LockingStatementOperations. Play nice.

 Please enter the commit message for your changes. Lines starting

Commit:
Play nice

tinwelint committed:
Removed no-locking option

tinwelint committed:
Preparation for resourceIds...

tinwelint committed:
A Test

tinwelint committed:
Commit context should have real locking

tinwelint committed:
Master conversations use locks delegate

tinwelint committed:
ClientAdapter

tinwelint committed:
Some lock batching

tinwelint committed:
Moar locking in the test

tinwelint committed:
Unified add/change property at commit

Transaction state before commit time represents logical changes that
are to be converted to physical changes during commit. This is true
for everything except property add/change. User can only "set"
properties, not differentiate between add or change by intent.
Yet the transaction state would capture and keep state for every
property whether or not it was an added or changed property,
i.e. whether or not the property already existed in the store at
the time of making the change in the transaction (state).
This information would then affect how the property changes would
be converted into physical changes, i.e. the logical transaction state
contained actual pysical state, which is fundamentally wrong.
This happened to work since write locks were acquired before reading
physical state affecting how a property change was represented
in the logical transaction state. With different locking strategies
this problem was exposed and so have to be fixed.

Now property changes will be treated simply as "set" operation,
i.e. "add or change" when converting logical state to physical
at commit time.

tinwelint committed:
Disables deferred locking by default

Commit:
Added missing test and documented missed termination case

tinwelint committed:
fixup! Prototype "finished", no tests, can maybe compile, no run, to be continued

tinwelint committed:
Uses community lock manager when deferring locks

Commit:
Added some tests for deferring locks

Commit:
Tests investigating deferring locks together with schema changes

Commit:
Set deferring-locks true by default

lutovich committed:
Turn off defered locks by default

To be able to build and verify that product works when they are turned
off.

lutovich committed:
Make SlaveLocksClient return correct delegate

Before this commit it was returning the underlying Locks as it's delegate.
This was wrong because delegate does not actually mean 'delegate'. Instead it
means client itself, if deferred locks are not used, and real underlying client,
if deferred locks are used.

lutovich committed:
Improve mocking in test

DefaultConversationSPITest mocks Locks. It failed when deferred locks were
enabled because mocks returned null. Fixed by RETURN_MOCKS answer.

Commit:
Refactor deferred locks to only defer where it is needed

Change perspective from "defer all locks and bypass where it is needed" to
"only defer lock where we actually want to".

This allow us to confine the logic of deferred locking to
KernelStatement, KernelTransactionImplementation, LockingStatementOperations.

It also allows for more conscious decisions to use implicit (deferred)
or explicit (not deferred) locking even if you are not sure if client is
wrapped by a deferring layer or not.

Commit:
Some clean up of DeferringLocksClient

Added some tests for DeferringLockClient
Clean up double release() in KTI

lutovich committed:
Make deferring lock client track reentrancy count

This is done to make sure sequence:
 1) acquireExclusive(1)
 2) acquireExclusive(1)
 3) releaseExclusive(1)
 4) acquireAllDeferredLocks()

results in one exclusive lock being grabbed for (1).

lutovich committed:
Add missing license headers

lutovich committed:
Enable deferred locking by default

Commit:
Add test for misuse of DeferringLockClient

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java:
```diff
@@ -78,14 +78,14 @@ void visit( ResourceType resourceType, long resourceId, String description, long
          * behavior is specified by the {@link WaitStrategy} for the given {@link ResourceType}.
          */
-        void acquireShared( ResourceType resourceType, long resourceId ) throws AcquireLockTimeoutException;
+        void acquireShared( ResourceType resourceType, long... resourceIds ) throws AcquireLockTimeoutException;
```
We need to add into java doc that those resources ids should be ordered in some way

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java:
```diff
@@ -256,36 +277,32 @@ private boolean getReadLockOnMaster( Locks.ResourceType resourceType, long resou
             }
             catch ( ComException e )
             {
                 throw new DistributedLockFailureException( "Cannot get shared lock on master", master, e );
```
Now we can have potentially multiple locks

burqen replied:
If we fail to grab lock, then the only sane thing to do is to rollback the transaction and retry it again. And rollback the transaction will release all locks.

MishaDemianenko replied:
Comment is about message in exception that is wrong now, sorry that I was not clear enought

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java:
```diff
@@ -256,36 +277,32 @@ private boolean getReadLockOnMaster( Locks.ResourceType resourceType, long resou
         }
         catch ( ComException e )
         {
             throw new DistributedLockFailureException( "Cannot get exclusive lock on master", master, e );
```
Now we can have potentially multiple locks

burqen replied:
Same as above.

MishaDemianenko replied:
Same as above

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java:
```diff
@@ -95,54 +97,75 @@ public SlaveLocksClient(
+            acquireSharedOnMaster( resourceType, newResourceIds );
+            for ( long resourceId : newResourceIds )
+            {
+                if ( client.trySharedLock( resourceType, resourceId ) )
```
We already teach lock clients to acquire multiple resources and it looks like it make sense to update 'try' calls to do that as well. WDYT?

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java:
```diff
@@ -95,54 +97,75 @@ public SlaveLocksClient(
+        for ( int i = 0; i < resourceIds.length; i++ )
         {
-            if ( client.trySharedLock( resourceType, resourceId ) )
+            assertNotStopped();
```
i do not think it make a lot of sense to check stop flag that often. Imho we can remove it completely here.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java:
```diff
@@ -78,14 +78,14 @@ void visit( ResourceType resourceType, long resourceId, String description, long
          * Can be grabbed when there are no locks or only share locks on a resource. If the lock cannot be acquired,
```
Now `Locks` interface looks weird: we allow acquisition of multiple resources but do not allow released them together.

tinwelint replied:
Right although there's no use case for releasing multiple afaik. The reason this was changed was so that we could implement the preparation of the locks, i.e. acquiring the deferred locks really optimized (and potentially do this in other places as well).

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/graphdb/Lock.java:
```diff
@@ -39,4 +39,12 @@
      */
     void release();
+
+    Lock NO_LOCK = new Lock()
```
not used

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/ConstraintEnforcingEntityOperations.java:
```diff
@@ -328,7 +328,7 @@ public long nodeGetFromUniqueIndexSeek(
 
         // If we find the node - hold a shared lock. If we don't find a node - hold an exclusive lock.
-        Locks.Client locks = state.locks();
+        Locks.Client locks = state.locks().implicit();
```
it looks like implicit locks, as they are defined now, are not suitable for this case

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -343,12 +351,17 @@ public KernelStatement acquireStatement()
         return currentStatement;
     }
 
+    private Locks.Client deferringLocks( Locks.Client locks )
```
never used

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -163,12 +167,13 @@ TransactionType upgradeToSchemaTransaction() throws InvalidTransactionTypeKernel
     private final Collection<Command> extractedCommands = new ArrayCollection<>( 32 );
     private final Locks locksManager;
     private final boolean txTerminationAwareLocks;
+    private final boolean deferringLocks;
```
we can completely avoid knowledge about deferring locks in `KernelTransactionImplementation` if we will pass instance of statementLocks that was constructed somewhere else.

burqen replied:
We could do it the same way we do with `locksManager`.
`this.statementLocks = statementLocksManager.newClient()`
and hide `locksManager` in `statementLocksManager`

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/CommunityEditionModule.java:
```diff
@@ -223,7 +223,13 @@ protected IdGeneratorFactory createIdGeneratorFactory( FileSystemAbstraction fs,
     public static Locks createLockManager( Config config, LogService logging )
     {
-        String key = config.get( GraphDatabaseFacadeFactory.Configuration.lock_manager );
+        // If we want to use deferred locking then use the community lock manager since it won't give
```
i do not think we should always stick to community lock manager and this comment is misleading

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacadeFactory.java:
```diff
@@ -87,8 +87,10 @@
                 " or `200k txs` for limiting the number of transactions to keep to 200 000", matches( ANY ) ) );
 
-        // Kept here to have it not be publicly documented.
+        // Kept here to have it not be publicly documented. community none
```
should not be here (new commend addition)

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/deferred/DeferringLockClient.java:
```diff
@@ -0,0 +1,198 @@
+        int cursor = 0;
+        Locks.ResourceType currentType = null;
+        boolean currentExclusive = false;
+        for ( LockUnit lockUnit : locks.keySet() )
```
what if we will split common locks into separate maps/arrays per resource type then we can significantly simplify acquisition and flushing. Also we will have more compact representation structure?

burqen replied:
Splitting the locks per type would make release and acquire more complex and we do not see how splitting will allow a more compact representation because we would still need to have some map structure to make release work efficient.

MishaDemianenko replied:
take a look into Forseti client

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/deferred/DeferringLockClient.java:
```diff
@@ -0,0 +1,198 @@
+public class DeferringLockClient implements Locks.Client
+{
+    private final Locks.Client clientDelegate;
+    private final Map<LockUnit,MutableInt> locks = new TreeMap<>();
```
what if we will sort only during acquisition instead of doing that always ?

burqen replied:
We need a map to make release efficient and if we use some unsorted map then we will have to copy the key set to sort it.

MishaDemianenko replied:
not really you do not need maps to make is sorted in the end, you can also take a look for example for forseti client

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/deferred/DeferringLockClient.java:
```diff
@@ -0,0 +1,198 @@
+    {
+        assertNotStopped();
+
+        if ( cursor > 0 )
```
looks like this check should be outside of this method

burqen replied:
Then we will have to make this check in two places. Having it inside the method avoids the duplication.

MishaDemianenko replied:
why do you need to have it in both places?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/StatementLocks.java:
```diff
@@ -0,0 +1,64 @@
+     *
+     * @return the locks client to serve explicit locks.
+     */
+    Locks.Client explicit();
```
really good that we split 2 type of locks but i still find naming a bit confusing. Current schema do not really represent nature for different type of locks and do not give good understanding when which type should be used.

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/locking/DeferringLocksIT.java:
```diff
@@ -0,0 +1,328 @@
+import static org.junit.Assert.assertTrue;
+import static org.neo4j.helpers.collection.Iterables.count;
+
+public class DeferringLocksIT
```
tests in this class should be repeated couple of times

burqen replied:
Fixed only test that was not deterministic.

Commit:
Remove unused code

Commit:
Knowledge about deferred locking abstracted away from KernelTransactionImplementation

... by introducing StatementLocksManager

Commit:
Addressing some PR comments

Added java doc to cover vararg on resourceIds in Locks.
Removed unnecessary assertNotStopped in SlaveLocksClient.

Commit:
Now grabbing explicit locks on SCHEMA and INDEX_ENTRY

This should protect from dirty reads of the schema store and from possible
duplicates in uniqueness constraints when using deferred locks.

Commit:
More descriptive naming for StatementLocks

explicit -> pessimistic because they are grabbed immediately
implicit -> optimistic because they can be deferred

Commit:
Clean up comments and exception message

lutovich committed:
Move deferred locking to a separate module and service

New module 'enterprise/deferred-locks' is introduced and components from it
are loaded via service loader. Jar obtained from this module is also not
included in packaging.

lutovich committed:
Take optimistic locks for SCHEMA and INDEX_ENTRY

This is done to make sure performance is not affected. Additional transaction
coordination on the righer level is required to handle schema and index
changes.

lutovich commented:
@tinwelint, @MishaDemianenko all comments addressed, PR is ready for the next review round.

**Update:** except `LockUnit` ordering, this we should discuss.

MishaDemianenko commented on enterprise/deferred-locks/src/main/java/org/neo4j/kernel/impl/locking/DeferringLockClient.java:
```diff
@@ -0,0 +1,194 @@
+import java.util.Map;
+import java.util.TreeMap;
+
+// TODO the state keeping in this class is quite unoptimized, please do so
```
please remove this todo as we agreed

MishaDemianenko commented on enterprise/deferred-locks/src/main/java/org/neo4j/kernel/impl/locking/DeferringLockClient.java:
```diff
@@ -0,0 +1,194 @@
+        boolean currentExclusive = false;
+        for ( LockUnit lockUnit : locks.keySet() )
+        {
+            // TODO perhaps also add a condition which sends batches over a certain size threshold
```
this as well

MishaDemianenko commented on enterprise/deferred-locks/src/main/java/org/neo4j/kernel/impl/locking/DeferringLockClient.java:
```diff
@@ -0,0 +1,194 @@
+
+    private void addLock( Locks.ResourceType resourceType, long resourceId, boolean exclusive )
+    {
+        assertNotStopped();
```
this check can be moved outside of this method

lutovich replied:
Method `addLock` is called from two places: `acquireShared` and `acquireExclusive`. This check will than be duplicated which is not nice.

MishaDemianenko replied:
well now that check is called in a loop - i do not think we should check that volatile that often

MishaDemianenko commented on enterprise/deferred-locks/src/main/java/org/neo4j/kernel/impl/locking/DeferringLockClient.java:
```diff
@@ -0,0 +1,194 @@
+
+    private void removeLock( Locks.ResourceType resourceType, long resourceId, boolean exclusive )
+    {
+        assertNotStopped();
```
this also can be moved outside

lutovich replied:
`removeLock` is called from `releaseShared` and `releaseExclusive`, probably better to check here and not duplicate as well.

MishaDemianenko replied:
well now that check is called in a loop - i do not think we should check that volatile that often

lutovich committed:
Removed unneeded TODOs and antrun plugin

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -250,9 +247,10 @@ public KernelTransactionImplementation( StatementOperationParts operations,
      */
-    public KernelTransactionImplementation initialize( long lastCommittedTx, long lastTimeStamp )
+    public KernelTransactionImplementation initialize( long lastCommittedTx, long lastTimeStamp,
+            StatementLocks statementLocks )
```
i think we should not pass `statementLocks` here, we should pass all required params through constructor and build them ourselves

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -600,7 +598,8 @@ private void commit() throws TransactionFailureException
                 }
 
-                context.init( locks );
+                statementLocks.prepareForCommit();
```
Sequence of operation is not really clear here, java docs does not anything about optimistic locks that will be taken here or why we init context with pessimistic lock only

MishaDemianenko commented:
@lutovich review round completed

lutovich committed:
Fix neo4j-deferred-locks module version after rebase

lutovich committed:
Make KTI create StatementLocks

KTI is not responsible for StatementLocks creation during `#initialize()`.
StatementLocks are not passed in as a parameter.

Commit also improves documentation around StatementLocks.

MishaDemianenko commented:
@lutovich do not have any open questions anymore, we have talked or resolve all open questions so far.
@tinwelint do you have any open comments ? LGTM?

lutovich committed:
Improve stop checks in DeferringLockClient

tinwelint commented:
@MishaDemianenko I'll promptly make a final review

lutovich committed:
Acquire deferred exclusive locks first

This is done to minimize amount of potential deadlocks which can happen if
transactions take their shared locks and then contend on same exclusive locks.

lutovich commented:
Changed lock ordering so that deferred exclusive locks are acquired first. Pr should be ready for merge now.

tinwelint commented:
@lutovich Changes look good. One thing though: it looks like (also shown in my local testing) that simply having the deferred locks jar on your classpath will always have deferred locking enabled. I don't think that's desirable. We should still have a config for enabling it, don't you think?

lutovich commented:
@tinwelint sounds good. i'll add this config

tinwelint commented:
@lutovich Another thing which was campaigned by @burqen was the lock order. There are two things we may consider here:
- `SHARED`+`EXCLUSIVE` on any given entity => `EXCLUSIVE`, i.e. `EXCLUSIVE` overwrites `SHARED` on an entity
- If order is decided using: `Resource type`, `ID`, `Lock type (i.e. exclusive/shared)` then locking will be free from deadlocks as a whole.

Implementing the new lock order would potentially result in more network calls when committing, but it may be worth it since the number of network calls are already heavily reduced compared to acquiring one by one and fewer retries of transactions will be needed.

lutovich committed:
Add config to enable/disable deferred locks

Also simplified StatementLocks creation in KTI - now it only requires
StatementLocksFactory and does not require Locks.

lutovich commented:
It was decided not to change lock ordering now.
@tinwelint, @MishaDemianenko added a commit with config and simplified locks creation. please review.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/SimpleStatementLocksFactory.java:
```diff
@@ -19,11 +19,36 @@
 {
+    private final Locks locks;
+
+    public SimpleStatementLocksFactory( Locks locks )
```
Initialisation lifecycle of `SimpleStatementLocksFactory` is a bit different then any other `StatementLocksFactory` which is a bit confusing. Can we please add default constructor and assign locks in `initialize` instead to much expected service behaviour, and mark `public SimpleStatementLocksFactory( Locks locks )` as test only for convenient test usage

MishaDemianenko commented:
@lutovich done, one small comment here and waiting for agreed setting description, please ping when done

lutovich committed:
Unify lifecycle of StatementLocksFactory implementations

And improve setting description.
